src/SMT.hs

```diff
@@ -59,36 +82,75 @@ data SMTExp = SMTExp
+      storage = text ";STORAGE:" <$$> (vsep . (fmap text) . nubOrd . _storage $ e) <> line
+      calldata = text ";CALLDATA:" <$$> (vsep . (fmap text) . nubOrd . _calldata $ e) <> line
+      environment = text ";ENVIRONMENT" <$$> (vsep . (fmap text) . nubOrd . _environment $ e) <> line
+      assertions = text ";ASSERTIONS:" <$$> (vsep . (fmap text) . nubOrd . _assertions $ e) <> line
 
--- A Query is a structured representation of an SMT query for an individual
--- expression, along with the metadata needed to pretty print the result
+-- | A Query is a structured representation of an SMT query for an individual
+--   expression, along with the metadata needed to extract a model from a satisfiable query
 data Query
-  = Postcondition Claim (Exp Bool) SMTExp
-  | Inv Claim Invariant SMTExp
+  = Postcondition Transition (Exp Bool) SMTExp
+  | Inv Invariant (Constructor, SMTExp) [(Behaviour, SMTExp)]
   deriving (Show)
 
 data SMTResult
-  = Sat
+  = Sat Model
   | Unsat
   | Unknown
   | Error Int String
   deriving (Show)
 
+-- | An assignment of concrete values to symbolic variables structured in a way
+--   to allow for easy pretty printing. The LHS of each pair is the symbolic
+--   variable, and the RHS is the concrete value assigned to that variable in the
+--   counterexample
 data Model = Model
-  { _mstorage :: Map Id (MType, MType)
-  , _mcalldata :: Map Id MType
-  , _menvironment :: Map Id MType
+  { _mprestate :: [(StorageLocation, ReturnExp)]
+  , _mpoststate :: [(StorageLocation, ReturnExp)]
+  , _mcalldata :: (String, [(Decl, ReturnExp)])
+  , _menvironment :: [(EthEnv, ReturnExp)]
+  -- invariants always have access to the constructor context
+  , _minitargs :: [(Decl, ReturnExp)]
   }
   deriving (Show)
 
+instance Pretty Model where
+  pretty (Model prestate poststate (ifaceName, args) environment initargs) =
+    (underline . text $ "counterexample:") <$$> line
+      <> (indent 2
+        (    calldata'
+        <$$> ifExists environment (line <> environment' <> line)
+        <$$> storage
+        <$$> ifExists initargs (line <> initargs')
+        ))
+    where
+      calldata' = (text "calldata:") <$$> line <> (indent 2 $ formatSig ifaceName args)
+      environment' = (text "environment:") <$$> line <> (indent 2 . vsep $ fmap formatEnvironment environment)
+      storage = (text "storage:") <$$> (indent 2 . vsep $ [ifExists prestate (line <> prestate'), poststate'])
+      initargs' = (text "constructor arguments:") <$$> line <> (indent 2 $ formatSig "constructor" initargs)
+
+      prestate' = (text "prestate:") <$$> line <> (indent 2 . vsep $ fmap formatStorage prestate) <> line
+      poststate' = (text "poststate:") <$$> line <> (indent 2 . vsep $ fmap formatStorage poststate)
+
+      formatSig iface cd = (text iface) <> (encloseSep lparen rparen (text ", ") $ fmap formatCalldata cd)
+      formatCalldata ((Decl _ name), val) = text $ name <> " : " <> prettyReturnExp val
```

Comment:
I'd maybe go for `=` instead of `:` here. I associate `:` with typing

Reply:
I'd change it to `=` for environment and storage too, for the same reason.

Reply:
Oh and you also don't need the parentheses here 🙃: 

```suggestion
      formatCalldata (Decl _ name, val) = text $ name <> " : " <> prettyReturnExp val
```